implement stri_detect_regex() in stringi.java

psolaimani opened the pull request.

Commit:
implement stri_detect_regex()

akbertram commented:
What if pattern has zero length?
And are you sure using the JRE's regular expression implementation is better than using Renjin's? I'm not sure about the syntax used by stringi, but Renjin's implementation is inching closer to PCRE syntax...

Alain-Bearez commented:
Using my current translation idiom, based on code found in #3, the code would read as follow:

```java
  public static SEXP stri_detect_regex(SEXP str, SEXP pattern, SEXP negate, SEXP opts_regex) { 
    final boolean is_negating = ((AtomicVector) negate).getElementAsLogical(0).toBooleanStrict();
    final int flags = __regex_flags(opts_regex);
    final int length = __recycling_rule(true, str, pattern);
    final Logical[] result = new Logical[length];
    final StringVector strings = __ensure_length(length, stri_prepare_arg_string(str, "str"));
    final StringVector patterns = __ensure_length(length, stri_prepare_arg_string(pattern, "pattern"));

    for (int i = 0; i < length; i++) {
      if (strings.isElementNA(i) || patterns.isElementNA(i) || patterns.getElementAsString(i).length() <= 0) {
        if (!patterns.isElementNA(i) && patterns.getElementAsString(i).length() <= 0) {
          Native.currentContext().warn("empty search patterns are not supported");
        }
        result[i] = Logical.NA;
      } else {
        final String element = strings.getElementAsString(i);
        final String appliedPattern = __normalize_binary_properties(patterns.getElementAsString(i));
        final Matcher matcher = Pattern.compile(appliedPattern, flags).matcher(element);
        final boolean found = matcher.find();
        result[i] = Logical.valueOf(is_negating ? !found : found);
      }
    }
    
    return new LogicalArrayVector(result);
  }
```

Of course, we would need to implement other `stri_detect_` functions for the examples to pass.

Commit:
improve implementation stri_detect_regex()

Commit:
merge woth bdd fork

Commit:
handle empty strings or patterns

example:
> stri_detect_regex(c('aaa','abc','NaN','111','123','NA'), c())
logical(0)

Commit:
handle LogicalVector strings and patterns

example:
> stri_detect_regex(c(TRUE,TRUE,TRUE), c('a','T',TRUE))
[1] FALSE  TRUE  TRUE

psolaimani commented:
@Alain-Bearez thanks for the implementation, added two edge cases ( empty argument `c()` and logical vector in input.
